### PR TITLE
Add basic(-ally all) tests

### LINKED ISSUE #20: Add tests *now* - start with frontend
I've thought about rewriting (more like copy-pasting together) the whole thing with TDD.
Maybe that's a bit too extreme, will have to time how long part of that takes and estimate.
One nice effect this would have is better commit messages, since early history is missing in the current version.
If it's not helpful I can just take the tests I will have written back to the current version.

### PR DESCRIPTION
**Pull request by hecatia-elegua**
Closes #20.
@pickx take a look, did I miss some basic thing we should include?

- [x] I'll move over the nested examples later and also add `array_at` testing.
- [x] Still have to update the README.
- [x] Also reminds me that I should update the `fallback` stuff in the README.
- [x] Also-Also some quick fixes found while testing

### CODE REVIEW

File: tests/ui/attr-value-is-invalid.rs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+// #[bitsize(0)]
+// enum Test {}
+
+// one above highest (struct) value
+#[bitsize(129)]
+struct Test {}
+
+// one above highest enum value
+#[bitsize(65)]
+enum Test {}
```
Comment (pickx) [inline tests/ui/attr-value-is-invalid.rs:17]: now with #45, this fails with `empty enums are not supported`.

worth noting that the empty struct above it _does_ fail with the expected message. and the `#[bitsize("12")]` test below also fails as expected.
so we should probably change the order of checks in `crate::bitsize::analyze_enum`

File: bilge-impl/src/bitsize_internal.rs [hunk 2]
```diff
@@ -71,7 +71,7 @@ fn generate_struct(struct_data: &ItemStruct, arb_int: &TokenStream) -> TokenStre
         }
         impl #ident {
             // #[inline]
-            #[allow(clippy::too_many_arguments, clippy::type_complexity)]
+            #[allow(clippy::too_many_arguments, clippy::type_complexity, clippy::unused_parens)]
```
Comment (pickx) [inline bilge-impl/src/bitsize_internal.rs:74]:
```suggestion
            #[allow(clippy::too_many_arguments, clippy::type_complexity, unused_parens)]
```
this is a `rustc` lint and is not suppressed with the `clippy` qualifier

File: tests/ui/discriminant-invalid.stderr [hunk 3]
```diff
@@ -11,3 +11,21 @@ error: Value of variant exceeds the given number of bits
...
    |     ^^^^^^^^^^^^
+
+error: enum overflows its bitsize
+  --> tests/ui/discriminant-invalid.rs:22:10
+   |
+22 | #[derive(FromBits)]
+   |          ^^^^^^^^
+   |
+   = help: there should only be at most 2 variants defined
+   = note: this error originates in the derive macro `FromBits` (in Nightly builds, run with -Z macro-backtrace for more info)
```
Comment (pickx) [inline tests/ui/discriminant-invalid.stderr:22]: I like this message a lot

**Review by pickx — APPROVED** [review]

File: bilge-impl/src/bitsize.rs [hunk 4]
```diff
@@ -120,34 +120,28 @@ fn split_attributes(item: &Item) -> SplitAttributes {
...
-                                    } else {
-                                        // It is most probable that basic derive macros work if we put them on after compression
-                                        after_compression.push(derive);
-                                    }
+                            // NOTE: we could also handle `::{path}`
+                            match path {
+                                "FromBytes" | "zerocopy :: FromBytes" => from_bytes = Some(derive_path),
+                                "FromBits" => has_frombits = true,
+                                "Debug" | "fmt :: Debug" | "core :: fmt :: Debug" | "std :: fmt :: Debug" => {
+                                    abort!(derive_path, "use derive(DebugBits) for structs")
```
Comment (pickx) [inline bilge-impl/src/bitsize.rs:132]: is this particular expansion of whitespace guaranteed here? (could do `.replace(' ', "")` if unsure)

File: tests/ui/attr-value-is-invalid.rs [hunk 5]
```diff
@@ -0,0 +1,23 @@
+use bilge::prelude::*;
+
+// negative value
+#[bitsize(-1)]
+enum Test {}
+
```
Comment (pickx) [inline tests/ui/attr-value-is-invalid.rs:6]: might want to check `0` as well, since we used to allow it

File: bilge-impl/src/bitsize.rs (hunk 4 above)
Comment (hecatia-elegua) [reply bilge-impl/src/bitsize.rs:132]: yeah the proc macro internals and quote and so on seem to all do this, if it bothers us later we can use replace there and handle `::core::fmt::Debug` and so on while we're at it.

File: tests/ui/attr-value-is-invalid.rs (hunk 5 above)
Comment (hecatia-elegua) [reply tests/ui/attr-value-is-invalid.rs:6]: it's commented out because I started when the change wasn't merged 👍

**Review by pickx — APPROVED** [review]

**Pull request merged**